modules/c++/nitf/include/nitf/CppImageSubheader.hpp

```diff
@@ -0,0 +1,683 @@
+     * \return type of corners
+     */
+    CornersType getCornersType() const;
+
+    /*!
+     *  This function allows the user to extract corner coordinates as a
+     *  set of decimal values.  This function only supports CornersTypes of
+     *  NITF_GEO or NITF_DECIMAL.  Others will throw an exception.
+     *
+     *  The output corners will be oriented to correspond to
+     *  corners[0] = (0, 0),
+     *  corners[1] = (0, MaxCol),
+     *  corners[2] = (MaxRow, MaxCol)
+     *  corners[3] = (MaxRow, 0)
+     *
+     *  following in line with 2500C.
+     *  \return corner coordinates as a set of decimal values
+     */
+    types::LatLonCorners getCornersAsLatLons() const;
+
+    /*!
+     * Set the image dimensions and blocking info.
+     * The user specifies the number of rows and columns in the image, number
+     * of rows and columns per block, and blocking mode. The number of blocks
+     * per row and column is calculated. The NITF 2500C large block option can
+     * be selected for either dimension by setting the corresponding block
+     * dimension to 0.
+     *
+     * \param numRows           The number of rows
+     * \param numCols           The number of columns
+     * \param numRowsPerBlock   The number of rows/block
+     * \param numColsPerBlock   The number of columns/block
+     * \param imageMode         Image mode
+     */
+    void setBlocking(Uint32 numRows,
+            Uint32 numCols,
+            Uint32 numRowsPerBlock,
+            Uint32 numColsPerBlock,
+            const std::string& imageMode);
+
+    /*!
+     * Set the image dimensions and blocking.
+     *
+     * The blocking is set to the simplest possible blocking given the
+     * dimension, no blocking if possible. Blocking mode is set to B. Whether
+     * or not blocking masks are generated is dependent on the compression
+     * code at the time the image is written. NITF 2500C large block
+     * dimensions are not set by this function. The more general set blocking
+     * function must be used to specify this blocking option.
+     *
+     * \param numRows           The number of rows
+     * \param numCols           The number of columns
+     */
+    void setDimensions(Uint32 numRows, Uint32 numCols);
+
+    //! Get the number of bands
+    Uint32 getBandCount() const;
+
+    /*!
+     * Creates the specified number of bandInfo objects, and updates the
+     * bandInfo array. This also updates the NBANDS/XBANDS fields in the
+     * ImageSubheader.
+     */
+    void createBands(Uint32 bandCount);
+
+    //! Insert the given comment at the given index (zero-indexed);
+    void insertImageComment(const std::string& comment, size_t position);
+
+    //! Remove the given comment at the given index (zero-indexed);
+    void removeImageComment(size_t position);
+
+    inline Uint32 getNumImageComments() const
+    {
+        return mNumImageComments;
+    }
+
+    inline const std::vector<BCSA<std::string, 80> >& getImageComments() const
```

Comment:
So I have to think about this a little more but I think I really like the BCSA and BCSN classes you put together.  It makes the sizes really explicit and I'm thinking should make reading and writing easy.  I want to make sure the interface seems easy though... so in this case, I think returning a vector of strings (even though this'll make a copy - these'll be trivially small) is easier for the user to work with.